SIL: make a linkage mismatch during de-serialization a human readable diagnostic

eeckstein opened the pull request:
Although this error can only happen when tinkering with the stdlib's runtime functions, it's nice to get a readable error message. So far, the compiler just crashed later without a hint to the original problem.

eeckstein commented:
@swift-ci smoke test

slavapestov commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -142,6 +142,10 @@ ERROR(deserialize_function_type_mismatch,Fatal,
       (StringRef, Type, Type))
 
+ERROR(deserialize_function_linkage_mismatch,Fatal,
+      "linkage mismatch of function '%0', declared as %1but expected as %2",
```
I think you want to add a space between then `%1` and `but`

slavapestov reviewed (commented):
I won't insist on a test case since I'm assuming it's too much of a pain in the ass, but if it's at all possible...

rauhul commented:
Thanks for improving this erik! :D

eeckstein committed:
SIL: make `getLinkageString(SILLinkage linkage)` available as general API

eeckstein committed:
SIL: make a linkage mismatch during de-serialization a human readable diagnostic

Although this error can only happen when tinkering with the stdlib's runtime functions, it's nice to get a readable error message. So far, the compiler just crashed later without a hint to the original problem.

slavapestov commented:
Thank you for addressing the review feedback! LGTM.
